### ISSUE TITLE
Need clarity around GMP with managed collection

### ISSUE THREAD

**Issue by ShibraAmin18** [GoogleCloudPlatform/prometheus-engine#226]
It will be nice to get documentation around setting Google Managed Prometheus with whole GKE monitoring (cluster, nodes, workloads, pods, and other resources) as we get in the Kube-prometheus-stack chart. 

I enabled managed prometheus in GKE and I can verify that collectors are running as daemonset and rule-evaluator, gmp-operator are running as deployment. 

How do we set up Podmonitoring or ClusterPodMonitoring for monitoring all kinds of resources ?

Do we need node exporter and Kube state metrics in GMP also?

How can we change prometheus.yaml config to add different exporter job scrape config ?

Thanks.

**Comment by bespokeeagle** [comment]
@ShibraAmin18 i found this https://cloud.google.com/stackdriver/docs/managed-prometheus/setup-managed helpful concerning your questions around PodMonitoring and ClusterPodMonitoring.

i am also curious about the kube state metrics, i currently have kube state metrics configured for my managed prometheus though but i am wondering if it is better to make use of kubelet and Cadvisor because it's in the docs.

You can configure a PodMonotioring CR for kube state metrics.

**Comment by lyanco** [comment]
We are working on publishing copy-pastable scrape configs that would support the exporters installed by kube-prometheus. The first instructions will be for kube-state, and more will come as time goes on.

You shouldn't have to modify prometheus.yaml if using managed collection - most everything can be scraped by a PodMonitoring or ClusterPodMonitoring.

**Comment by ShibraAmin18** [comment]
@lyanco So for now, managed collection can work with exporters or not?

**Comment by bespokeeagle** [comment]
> @lyanco So for now, managed collection can work with exporters or not?
> 
> I'm working with node exporter and created PodMonitoring for the same. But not getting any metrics in GMP.

i had the same issue but with kube state metrics, after i added a section for it in the scrape config for the prometheus i started getting the metrics

**Comment by lyanco** [comment]
I think there's a bit of a mental model mismatch here. The collector **is** (slightly modified) Prometheus. Managed collection is our set of conventions to scalably deploy and configure Prometheus.

You can use a PodMonitoring or ClusterPodMonitoring resource to configure managed Prometheus to scrape any exporter. We're working on making this more debuggable, out-of-the-box, and managed, but for now you'd have to install exporters and configure scraping for the exporters yourself.

See https://cloud.google.com/stackdriver/docs/managed-prometheus/setup-managed

**Comment by bespokeeagle** [comment]
@lyanco thanks for clearing that up.
@ShibraAmin18  so sorry to have hijacked your issue, as @lyanco said the PodMonoting or ClusterPodMonitoring can scrape any exporter.
Things to note.
1. PodMonitoring CR should be configured in the same namespace as exporter you want to scrape. You can create as many as needed
2. Ensure to use correct port name in your PodMonitoring manifest.
3. Double check your labels

**Comment by elja** [comment]
Recently we migrated GKE workload metrics to GMP. However, I can't find a clear answer on if it's possible to use GMP metrics in HPA (previously it just worked without any additional magic by the workload metric name in the HPA definition). At this point, it's not really clear if I need to export GMP metrics to the Kubernetes cluster with some stack driver adapter or exporter.

**Comment by lyanco** [comment]
Hi Alex, 

You're not the only one to have reported this - HPA should work. We'll look into it and will get back to you by end of next week, most likely.

**Comment by lyanco** [comment]
OK, well I have news. That was fast.

While the custom metrics adapter is still mysteriously not working, HPA using [prometheus-adapter](https://github.com/kubernetes-sigs/prometheus-adapter) should work with GMP now. The only caveat is the .seriesQuery field of the rules config needs to specify the metric name directly; it can't use a regex. You also need to deploy the [frontend](https://cloud.google.com/stackdriver/docs/managed-prometheus/query#ui-prometheus) and point the HPA adapter at it, e.g.
        - --prometheus-url=http://frontend.default.svc:9090/

Want to give it a shot and see if it works for you?

**Comment by shpml** [comment]
There seems to be some clashes with names when deploying [prometheus-adapter](https://github.com/kubernetes-sigs/prometheus-adapter/tree/master/deploy/manifests) along side [custom-metrics-stackdriver-adapter](https://github.com/GoogleCloudPlatform/k8s-stackdriver/tree/master/custom-metrics-stackdriver-adapter). They both deploy new API services. 

The below are some resources that needed to be changed back when running terraform apply against my `custom-metrics-stackdriver-adapter` terraform module. 
```
  ~ resource "kubernetes_api_service" "v1beta1-custom" {
        id = "v1beta1.custom.metrics.k8s.io"


      ~ spec {
            # (5 unchanged attributes hidden)

          ~ service {
              ~ name      = "custom-metrics-apiserver" -> "custom-metrics-stackdriver-adapter"
                # (2 unchanged attributes hidden)
            }
        }
        # (1 unchanged block hidden)
    }

  ~ resource "kubernetes_api_service" "v1beta1-external" {
        id = "v1beta1.external.metrics.k8s.io"


      ~ spec {
            # (5 unchanged attributes hidden)

          ~ service {
              ~ name      = "custom-metrics-apiserver" -> "custom-metrics-stackdriver-adapter"
                # (2 unchanged attributes hidden)
            }
        }
        # (1 unchanged block hidden)
    }

  ~ resource "kubernetes_api_service" "v1beta2-custom" {
        id = "v1beta2.custom.metrics.k8s.io"


      ~ spec {
            # (5 unchanged attributes hidden)

          ~ service {
              ~ name      = "custom-metrics-apiserver" -> "custom-metrics-stackdriver-adapter"
                # (2 unchanged attributes hidden)
            }
        }
        # (1 unchanged block hidden)
    }

  ~ resource "kubernetes_cluster_role_binding" "auth-delegator" {
        id = "custom-metrics:system:auth-delegator"



      ~ subject {
          ~ name      = "custom-metrics-apiserver" -> "custom-metrics-stackdriver-adapter"
            # (2 unchanged attributes hidden)
        }
        # (2 unchanged blocks hidden)
    }

  ~ resource "kubernetes_role_binding" "custom_metrics_auth_reader" {
        id = "kube-system/custom-metrics-auth-reader"



      ~ subject {
          ~ name      = "custom-metrics-apiserver" -> "custom-metrics-stackdriver-adapter"
            # (2 unchanged attributes hidden)
        }
        # (2 unchanged blocks hidden)
    }

Plan: 0 to add, 5 to change, 0 to destroy.
```

**Comment by lyanco** [comment]
@shpml Do you think these instructions below cover what you just wrote in a succinct way?

You can run`prometheus-adapter` alongside `custom-metrics-stackdriver-adapter`, but in
order to avoid naming collisions you have to change the `spec.service.name` and
`subject.name` fields within your Stackdriver Adapter resources from
`custom-metrics-apiserver` to `custom-metrics-stackdriver-adapter`.

If so, I'll add that to [this new documentation page on HPA](https://cloud.google.com/stackdriver/docs/managed-prometheus/hpa) and I think we're good to go.

**Comment by shpml** [comment]
I don't think it's just a naming collision, I decided to delete `custom-metrics-stackdriver-adapter` resources in order to test `prometheus-adapter`. Although I would rather stay with `custom-metrics-stackdriver-adapter` since the config for `prometheus-adapter` looks like too much work to maintain. I'd rater not maintain a set of "discovery" rules that need to be updated when/if I expose new metrics.

They both deploy APIService resources with the same name [here](https://github.com/GoogleCloudPlatform/k8s-stackdriver/blob/d0679be2f5a723e08df99a857e691e5b39461515/custom-metrics-stackdriver-adapter/deploy/production/adapter.yaml#L108) and [here](https://github.com/kubernetes-sigs/prometheus-adapter/blob/master/deploy/manifests/custom-metrics-apiservice.yaml), `v1beta1.custom.metrics.k8s.io`, `v1beta2.custom.metrics.k8s.io` and `v1beta1.external.metrics.k8s.io`. And they both associate these APIService resources with different Service resources, `custom-metrics-apiserver` and `custom-metrics-stackdriver-adapter`

The `prometheus-adapter` service sends traffic to pods with the label [app = custom-metrics-apiserver](https://github.com/kubernetes-sigs/prometheus-adapter/blob/master/deploy/manifests/custom-metrics-apiserver-service.yaml) but the `custom-metrics-stackdriver-adapter` sends traffic to pods with labels [run and k8s-app = custom-metrics-stackdriver-adapter](https://github.com/GoogleCloudPlatform/k8s-stackdriver/blob/d0679be2f5a723e08df99a857e691e5b39461515/custom-metrics-stackdriver-adapter/deploy/production/adapter.yaml#L104)

It seems to me like they both can't be run in parallel due to the names of the APIService resources being the same, `v1beta1.custom.metrics.k8s.io`, `v1beta2.custom.metrics.k8s.io` and `v1beta1.external.metrics.k8s.io`. I'm not across the details on deploying kubernetes custom metrics and external metrics APIs but at a guess these are values kubernetes is expecting.

TL;DR; I think you can only deploy one implementation of kubernetes custom metrics and external metrics APIs.

**Comment by lyanco** [comment]
Re: the original thread, we've added copy-pastable YAML for common exporters and will continue to update this page: https://cloud.google.com/stackdriver/docs/managed-prometheus/exporters

**Comment by lyanco** [comment]
FYI we just released support for `custom-metrics-stackdriver-adapter`: https://cloud.google.com/stackdriver/docs/managed-prometheus/hpa
